www/commandQueueExecutor.js

```diff
@@ -155,21 +131,20 @@ function _exec() {
     // If the `_stopRequested` flag is true,
     // do not execute any statements except `remove()` or `clear()` methods.
     if (_stopRequested && (!commandParams.execOptions.remove || methodName !== 'clear')) {
-      _executingCnt--;
       continue;
     }
 
     // Some methods have to block other execution requests, such as `map.clear()`
     if (commandParams.execOptions.sync) {
-      _isWaitMethod = methodName;
       cordova_exec.apply(this, commandParams.args);
       break;
     }
     cordova_exec.apply(this, commandParams.args);
   }
 
-  _isExecuting = false;
-
+  setTimeout(function() {
+    common.nextTick(_exec);
+  }, 50);
```

Comment:
What's the rationale for choosing `50` over `0`?

If you just want to make to asynchronous and make it execute on the next tick, then using `0` is sufficient to make it execute later after this callstack but as early as possible.

`common.nextTick` might be already doing this too.

Reply:
var nextTick = resolvedPromise ? function (fn) {
  resolvedPromise.then(fn);
} : function (fn) {
  setTimeout(fn);
};

Seems like nextTick uses a default value, which is 0.

The 50 millisecond timeout delays the javascript a small amount to give time for native cordova to process the batch of requests that were just sent (max 100 per batch).

The previous code accomplished similar behaviour as it would call _exec on the success/fail callbacks of each item and then return early out of _exec if it was already running.

However, the previous code was messy. It had a lot of unnecessary _exec calls and depended on native APIs to actually return, a detail that I wouldn't trust in most cordova plugins, unless it was written by us.

The 50 millisecond timeout is more predictable and explicit and not prone to the side effects of unknown native details.

Reply:
This just occurred to me but... In the previous code, if the last call to be pushed never returned... I think that would halt the entire queue permanently.

I forget the exact condition that was causing the APIs needed by our map overlay to not get executed, but it's possible this is exactly why...

Reply:
> The 50 millisecond timeout is more predictable and explicit and not prone to the side effects of unknown native details.

If having to wait ~50ms is significant then the whole thing is not really reliable as we cannot guarantee that the native tasks will be completed within 50ms.

Either the process neeeds to be not sensitive to task time, or there needs a proper callback system...

Reply:
I don't think it will be a huge deal personally. I would think most calls will be completed in a few milliseconds.
Most of the time, there shouldn't be a huge load of calls. If there was a huge load, it could start sending the next batch before the first batch is completed.

But persistent huge loads are abnormal and we aren't planning on trying to PR this particular change into the master branch (we'll keep it to totalpave/master).

In addition... A proper callback system will involve native changes and begin depending on the response from each API call, which I don't particularly trust to be bullet proof and guaranteed to always return.

Reply:
For the record, I have ran this on tracker and pci. I wouldn't say that I properly tested huge loads, but it does function fine and I did not notice any performance drops.

Reply:
Maybe I'm not understanding the context but anything that arbitrary waits is a pretty big red flag.

If it doesn't always work waiting 0ms, then it won't always work waiting 50ms either.

Reply:
If you overload the cordova bridge with API calls, apparently it might just ignore some of the calls.
This is why the commandQueue system was originally made.

That's part of the reason why it's 50ms instead of 0ms.

Reply:
That's not making much sense to me.

Is the intent to stagger API calls rather than to wait for an action?

Reply:
Yes, that's one way to describe it. 

It's staggering the batches of calls instead of throwing them at the cordova bridge as fast as it can.
It's also doing that without relying on individual calls to guarantee return with a success or fail.

Reply:
Ok, I still think this is faulty logic then. Like this isn't staggering, it's shifting...

Like consider these examples:

```javascript
// Example 1
for (let i = 0; i < 1000; i++) {
  common.nextTick(_exec);
}

// Example 2
for (let i = 0; i < 1000; i++) {
  window.setTimeout(() => {
    common.nextTick(_exec);
  }, 0);
}

// Example 3
for (let i = 0; i < 1000; i++) {
  window.setTimeout(() => {
    common.nextTick(_exec);
  }, 50);
}

// Example 4
for (let i = 0; i < 1000; i++) {
  window.setTimeout(() => {
    common.nextTick(_exec);
  }, 1000);
}
```

Every one of these all executes `_exec` at roughly the same time.
Example 1 executes all of them immediately.
Example 2 executes all of them immediately after the current callstack
Example 3 executes all of them roughly 50ms later
Example 4 executes all of them roughly 1000ms later.